pam-openshift/pam_openshift.8
.\"     Title: pam_openshift
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 01/13/2012
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_LIBRA" "8" "01/13/2012" "Linux-PAM Manual" "Linux\-PAM Manual"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
pam_openshift \- PAM module to set the default security context
.SH "Synopsis"
.fam C
.HP \w'\fBpam_openshift\&.so\fR\ 'u
\fBpam_openshift\&.so\fR [close] [debug] [open] [verbose]
.fam
.SH "DESCRIPTION"
.PP
In a nutshell, pam_openshift sets up the default security context for the next execed shell\&.
.PP
When an application opens a session using pam_openshift, the shell that gets executed will be run in the default security context
.PP
Adding pam_openshift into a pam file could cause other pam modules to change their behavior if the exec another application\&. The close and open option help mitigate this problem\&. close option will only cause the close portion of the pam_openshift to execute, and open will only cause the open portion to run\&. You can add pam_openshift to the config file twice\&. Add the pam_openshift close as the executes the open pass through the modules, pam_openshift open_session will happen last\&. When PAM executes the close pass through the modules pam_openshift close_session will happen first\&.
.SH "OPTIONS"
.PP
\fBclose\fR
.RS 4
Only execute the close_session portion of the module\&.
.RE
.PP
\fBdebug\fR
.RS 4
Turns on debugging via
\fBsyslog\fR(3)\&.
.RE
.PP
\fBopen\fR
.RS 4
Only execute the open_session portion of the module\&.
.RE
.PP
\fBverbose\fR
.RS 4
attempt to inform the user when security context is set\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
Only the
\fBsession\fR
module type is provided\&.
.SH "RETURN VALUES"
.PP
PAM_AUTH_ERR
.RS 4
Unable to get or set a valid context\&.
.RE
.PP
PAM_SUCCESS
.RS 4
The security context was set successfully\&.
.RE
.PP
PAM_USER_UNKNOWN
.RS 4
The user is not known to the system\&.
.RE
.SH "EXAMPLES"
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
.if t \{\
.sp -1
.\}
.BB lightgray adjust-for-leading-newline
.sp -1

auth     required  pam_unix\&.so
session  required  pam_permit\&.so
session  optional  pam_openshift\&.so
    
.EB lightgray adjust-for-leading-newline
.if t \{\
.sp 1
.\}
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.SH "SEE ALSO"
.PP

\fBpam.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_openshift was written by Dan Walsh <dwalsh@redhat\&.com>\&.


pam-openshift/ChangeLog
Fri Jan 13 2012 Dan Walsh <dwalsh@redhat.com>
- First public release.


stickshift/pam-openshift/pam_openshift.c
/******************************************************************************
 * A module for Linux-PAM that will set the default security context after login
 * via PAM.
 *
 * Copyright (c) 2012 Red Hat, Inc.
 * Written by Dan Walsh <dwalsh@redhat.com>
 * Additional improvements by Tomas Mraz <tmraz@redhat.com>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <errno.h>
#include <limits.h>
#include <pwd.h>
#include <grp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>

#define PAM_SM_AUTH
#define PAM_SM_SESSION

#include <security/pam_modules.h>
#include <security/_pam_macros.h>
#include <security/pam_modutil.h>
#include <security/pam_ext.h>

#include <selinux/selinux.h>
#include <selinux/selinux.h>
#include <selinux/context.h>
#include <selinux/get_default_type.h>

#include <attr/attributes.h>

#ifdef HAVE_LIBAUDIT
#include <libaudit.h>
#include <sys/select.h>
#include <errno.h>
#endif

/* Send audit message */
static

int send_audit_message(pam_handle_t *pamh, int success, security_context_t default_context,
		       security_context_t selected_context)
{
	int rc=0;
#ifdef HAVE_LIBAUDIT
	char *msg = NULL;
	int audit_fd = audit_open();
	security_context_t default_raw=NULL;
	security_context_t selected_raw=NULL;
	rc = -1;
	if (audit_fd < 0) {
		if (errno == EINVAL || errno == EPROTONOSUPPORT ||
                                        errno == EAFNOSUPPORT)
                        return 0; /* No audit support in kernel */
		pam_syslog(pamh, LOG_ERR, "Error connecting to audit system.");
		return rc;
	}
	if (selinux_trans_to_raw_context(default_context, &default_raw) < 0) {
		pam_syslog(pamh, LOG_ERR, "Error translating default context.");
		default_raw = NULL;
	}
	if (selinux_trans_to_raw_context(selected_context, &selected_raw) < 0) {
		pam_syslog(pamh, LOG_ERR, "Error translating selected context.");
		selected_raw = NULL;
	}
	if (asprintf(&msg, "pam: default-context=%s selected-context=%s",
		     default_raw ? default_raw : (default_context ? default_context : "?"),
		     selected_raw ? selected_raw : (selected_context ? selected_context : "?")) < 0) {
		pam_syslog(pamh, LOG_ERR, "Error allocating memory.");
		goto out;
	}
	if (audit_log_user_message(audit_fd, AUDIT_USER_ROLE_CHANGE,
				   msg, NULL, NULL, NULL, success) <= 0) {
		pam_syslog(pamh, LOG_ERR, "Error sending audit message.");
		goto out;
	}
	rc = 0;
      out:
	free(msg);
	freecon(default_raw);
	freecon(selected_raw);
	close(audit_fd);
#else
	pam_syslog(pamh, LOG_NOTICE, "pam: default-context=%s selected-context=%s success %d", default_context, selected_context, success);
#endif
	return rc;
}

static int
send_text (pam_handle_t *pamh, const char *text, int debug)
{
  if (debug)
    pam_syslog(pamh, LOG_NOTICE, "%s", text);
  return pam_info (pamh, "%s", text);
}

static security_context_t user_context=NULL;
static security_context_t prev_user_context=NULL;
static int selinux_enabled=0;

#define UNUSED __attribute__ ((unused))

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh UNUSED, int flags UNUSED,
		    int argc UNUSED, const char **argv UNUSED)
{
	/* Fail by default. */
	return PAM_AUTH_ERR;
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh UNUSED, int flags UNUSED,
	       int argc UNUSED, const char **argv UNUSED)
{
	return PAM_SUCCESS;
}

static void get_mcs_level(int uid, security_context_t *scon) {
	int SETSIZE = 1023;
	int TIER = SETSIZE;

	int ORD=uid;
	while ( ORD > TIER ) {
		ORD = ORD - TIER;
		TIER -= 1;
	}
	TIER = SETSIZE - TIER;
	ORD = ORD + TIER;
	asprintf(scon, "unconfined_u:system_r:openshift_t:s0:c%d,c%d", TIER, ORD);
}

/* checks if a user is on a list of members of the GID 0 group */
static int is_on_list(char * const *list, const char *member)
{
    while (list && *list) {
        if (strcmp(*list, member) == 0)
            return 1;
        list++;
    }
    return 0;
}

static int openshift_domain(pam_handle_t *pamh, struct passwd *pw) {
	struct group *grp;
        char *secontext;
        int selength;

	if (!pw->pw_uid) return 0;

        if (strlen(pw->pw_dir)!=0) {
          secontext=(char *)malloc(ATTR_MAX_VALUELEN);
          selength=ATTR_MAX_VALUELEN;
          if (attr_get(pw->pw_dir, "selinux",
                       secontext, &selength,
                       ATTR_DONTFOLLOW | ATTR_SECURE) == 0) {
            if(strncmp(secontext, "system_u:object_r:openshift_var_lib_t:s0",
                       selength)!=0) {
              free(secontext);
              return 0;
            }
          }
          free(secontext);
        }

	if ((grp = pam_modutil_getgrnam (pamh, "wheel")) == NULL) {
	    grp = pam_modutil_getgrgid (pamh, 0);
	} 
	if (!grp) return 1;
	if (pw->pw_gid == grp->gr_gid) return 0;
	if (!grp->gr_mem) return 1;
	return ! is_on_list(grp->gr_mem, pw->pw_name);
}

PAM_EXTERN int
pam_sm_open_session(pam_handle_t *pamh, int flags UNUSED,
		    int argc, const char **argv)
{
	int i, debug = 0;
	int verbose=0, close_session=0;
	int ret = 0;
	const char *username;
	const void *void_username;
	struct passwd *pw;
	
	/* Parse arguments. */
	for (i = 0; i < argc; i++) {
		if (strcmp(argv[i], "debug") == 0) {
			debug = 1;
		}
		if (strcmp(argv[i], "verbose") == 0) {
			verbose = 1;
		}
		if (strcmp(argv[i], "close") == 0) {
			close_session = 1;
		}
	}
	
	if (debug)
		pam_syslog(pamh, LOG_NOTICE, "Open Session");
	
	/* this module is only supposed to execute close_session */
	if (close_session)
		return PAM_SUCCESS;

	if (!(selinux_enabled = is_selinux_enabled()>0) )
		return PAM_SUCCESS;
	
	if (pam_get_item(pamh, PAM_USER, &void_username) != PAM_SUCCESS ||
	    void_username == NULL) {
		return PAM_USER_UNKNOWN;
	}
	username = void_username;
	
	pw = pam_modutil_getpwnam( pamh, username );
	if (!pw) {
		pam_syslog(pamh, LOG_ERR, "Unable to find uid for user %s", username);
		return -1;
	}

	if (openshift_domain(pamh, pw)) {
		get_mcs_level(pw->pw_uid, &user_context);
	} else {
		user_context = strdup("unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023");
	}
	if (!user_context) {
		pam_syslog(pamh, LOG_ERR, "Unable to generate User context for user %s: %s", username, strerror(errno));
		return -1;
	}
	
	if (getexeccon(&prev_user_context)<0) {
		prev_user_context=NULL;
	}
	send_audit_message(pamh, 1, user_context, user_context);
	ret = setexeccon(user_context);
	if (ret==0 && verbose) {
		char msg[PATH_MAX];
		snprintf(msg, sizeof(msg),
			 "Security Context %s Assigned", user_context);
		send_text(pamh, msg, debug);
	}
	if (ret) {
		pam_syslog(pamh, LOG_ERR,
			   "Error!  Unable to set %s executable context %s. %s",
			   username, user_context, strerror(errno));
		if (security_getenforce() == 1) {
			freecon(user_context);
			return PAM_AUTH_ERR;
		}
	} else {
		if (debug)
			pam_syslog(pamh, LOG_NOTICE, "set %s security context to %s",
				   username, user_context);
	}
	
	freecon(user_context);
	
	return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_close_session(pam_handle_t *pamh, int flags UNUSED,
		     int argc, const char **argv)
{
	int i, debug = 0, status = PAM_SUCCESS, open_session = 0;
	if (! (selinux_enabled ))
		return PAM_SUCCESS;
	
	/* Parse arguments. */
	for (i = 0; i < argc; i++) {
		if (strcmp(argv[i], "debug") == 0) {
			debug = 1;
		}
		if (strcmp(argv[i], "open") == 0) {
			open_session = 1;
		}
	}
	
	if (debug)
		pam_syslog(pamh, LOG_NOTICE, "Close Session");
	
	if (open_session)
		return PAM_SUCCESS;
	
	if (setexeccon(prev_user_context)) {
		pam_syslog(pamh, LOG_ERR, "Unable to restore executable context %s: %s",
			   prev_user_context ? prev_user_context : "", strerror(errno));
		if (security_getenforce() == 1)
			status = PAM_AUTH_ERR;
		else
			status = PAM_SUCCESS;
	} else if (debug)
		pam_syslog(pamh, LOG_NOTICE, "Executable context back to original");
	
	if (prev_user_context) {
		freecon(prev_user_context);
		prev_user_context = NULL;
	}
	
	return status;
}
